### ISSUE TITLE
[Misc] Lambda Task timed out

### ISSUE THREAD

**Issue by tanimura55** [awslabs/fhir-works-on-aws-deployment#825]
**What's on your mind?**

## Question about Lambda Timeout Issue

Dear FHIR Works on AWS develop Community,

We are experiencing a timeout issue with `fhirServer` Lambda function.

Here is an example of the log when the problem occurred,
```
2023-02-27T17:25:37.766+09:00 START RequestId: d808b5a1-dcc5-44a9-a5bc-826a2b84c541 Version: 1
2023-02-27T17:26:17.836+09:00 REPORT RequestId: d808b5a1-dcc5-44a9-a5bc-826a2b84c541 Duration: 40068.65 ms Billed Duration: 40000 ms Memory Size: 512 MB Max Memory Used: 332 MB XRAY TraceId: 1-63fc6901-22f57b575c04477409babc4a SegmentId: 3be74cbd70118271 Sampled: true
2023-02-27T17:26:17.836+09:00 END RequestId: d808b5a1-dcc5-44a9-a5bc-826a2b84c541
2023-02-27T17:26:17.836+09:00 2023-02-27T08:26:17.835Z d808b5a1-dcc5-44a9-a5bc-826a2b84c541 Task timed out after 40.07 seconds
```

We have already consulted with AWS Support and tried extending timeout time and enabling Lambda provisioned cocurrency, but these measures have not yielded any significant insights.
And, we have checked the Lambda function's logs, as well as the X-Ray service map, but we could not identify which part of the function is causing the timeout or whether there are any connection errors with other services.

X-ray screenshot:
![screencapture-x-ray](https://user-images.githubusercontent.com/98066703/224939268-35fdab8d-e8c3-40c6-ad28-401f4a3d4dff.png)

We would like to ask for your advice on how to further investigate this issue. 
Is there any specific tool or method for investigating?
Is there any known issue with FHIR Works on AWS that might be causing this problem?

Thank you.

**Versions (please complete the following information):**
 - [v4.2.0](https://github.com/awslabs/fhir-works-on-aws-deployment/releases/tag/v4.2.0)
 - commit https://github.com/awslabs/fhir-works-on-aws-deployment/commit/432d4a88e74c4bfec5b0374ee21f592aaebaef02

**Comment by Bingjiling** [comment]
Hi @tanimura55, 

Thanks for reaching out! We have not seen the issue described on our end. Here are some areas to look into that might help debugging: 

1. Is your system experiencing Lambda time out issue with all FWoA API endpoints? Have you tried the `/matadata` endpoint? 
2. Which version of FWoA is used? Were there any changes to the codes? 
3. Is CloudTrail turned on for the AWS account? Have you looked into the AWS Service calls from Fhir server Lambda in CloudTrial? 
4. Have you looked into the Lambda metrics on CPU and memory usage? 

Thanks,
Yanyu

**Comment by tanimura55** [comment]
Thank you for your quick response @Bingjiling !

Sorry for the lack of information.
This problem occurred when we tested the load of an application that has FWoA as its backend using `k6` to see how much load the stack on which FWoA was deployed can withstand.
Therefore, it is a special situation.

Regarding your proposal:

> 1. Is your system experiencing Lambda time out issue with all FWoA API endpoints? Have you tried the /matadata endpoint?

We only used the following endpoints for testing:
- `/Observation`
- `/Patient`

We did not use the `/metadata` endpoint.

> 2. Which version of FWoA is used? Were there any changes to the codes?

We used version is [v4.2.0](https://github.com/awslabs/fhir-works-on-aws-deployment/releases/tag/v4.2.0).
There were no changes made to the code.

> 3. Is CloudTrail turned on for the AWS account? Have you looked into the AWS Service calls from Fhir server Lambda in CloudTrial?

Although the request IDs are different, the following logs called from Lambda were confirmed by CloudTrail during the period when the timeout occurred.
It appears to be only API calls for DynamoDB decryption.

<details>
<summary>CloudTrail log</summary>

Details |  
-- | --
Event time | February 27, 2023, 17:27:00 (UTC+09:00)
User name | fhir-service-dev-fhirServer
Event name | Decrypt
Event source | http://kms.amazonaws.com/
Source IP address | http://dynamodb.amazonaws.com/
Event ID | xxx
Request ID | xxx
AWS region | ap-northeast-1
Error code | -
Read-only | true

</details>

Since we were not able to identify the event by identifying the request ID, this event may not have occurred in the corresponding Lambda execution.

> 4. Have you looked into the Lambda metrics on CPU and memory usage?

Since we could not reference the metrics that were executed on the 27th, we executed them again under the same conditions.
At that time, the Lambda function was set to `1024 MB` memory and `2 minutes timeout`.
The following metrics were output when the timeout occurred.

![image](https://user-images.githubusercontent.com/98066703/225565495-a895f35c-ebfe-43cf-9fa8-bbb33eb85512.png)

Looking at the graph again, I noticed that CPU utilization peaked out at the same time as the time the timeout occurred, so the CPU limit may be suspect.

And for the Lambda Request ID where the timeout was observed, the following metrics were obtained.

<details>
<summary>Lambda Insights log</summary>

Log content  | Detail
-- | --
@ingestionTime | 1678951551235
@log | xxx:/aws/lambda-insights
@logStream | fhir-service-dev-fhirServer/CuaItBE2KZzFkcTJ2nx3abZO9
@message | {"timeout":true,"memory_utilization":41,"tmp_free":538324992,"_aws":{"CloudWatchMetrics":[{"Namespace":"LambdaInsights","Dimensions":[["function_name"],["function_name","version"]],"Metrics":[{"Name":"cpu_total_time","Unit":"Milliseconds"},{"Name":"tx_bytes","Unit":"Bytes"},{"Name":"rx_bytes","Unit":"Bytes"},{"Name":"total_network","Unit":"Bytes"},{"Name":"tmp_used","Unit":"Bytes"},{"Name":"memory_utilization","Unit":"Percent"},{"Name":"total_memory","Unit":"Megabytes"},{"Name":"used_memory_max","Unit":"Megabytes"}]}],"Timestamp":1678951430387,"LambdaInsights":{"ShareTelemetry":true}},"fd_max":1024,"function_name":"fhir-service-dev-fhirServer","version":"provisioned","rx_bytes":1110,"billed_duration":120000,"fd_use":21,"request_id":"8fec99e8-fd41-458c-b8b6-c11f48c1b31e","agent_memory_avg":11,"cpu_total_time":1270,"tx_bytes":78195,"total_network":79305,"agent_version":"1.0.143.0","cpu_system_time":690,"event_type":"performance","tmp_used":12136448,"threads_max":10,"total_memory":1024,"cpu_user_time":580,"shutdown":1,"duration":120532,"cold_start":false,"agent_memory_max":11,"tmp_max":550461440,"shutdown_reason":"timeout","billed_mb_ms":122880000,"used_memory_max":421}
@timestamp | 1678951551151
_aws.CloudWatchMetrics.0.Dimensions.0.0 | function_name
_aws.CloudWatchMetrics.0.Dimensions.1.0 | function_name
_aws.CloudWatchMetrics.0.Dimensions.1.1 | version
_aws.CloudWatchMetrics.0.Metrics.0.Name | cpu_total_time
_aws.CloudWatchMetrics.0.Metrics.0.Unit | Milliseconds
_aws.CloudWatchMetrics.0.Metrics.1.Name | tx_bytes
_aws.CloudWatchMetrics.0.Metrics.1.Unit | Bytes
_aws.CloudWatchMetrics.0.Metrics.2.Name | rx_bytes
_aws.CloudWatchMetrics.0.Metrics.2.Unit | Bytes
_aws.CloudWatchMetrics.0.Metrics.3.Name | total_network
_aws.CloudWatchMetrics.0.Metrics.3.Unit | Bytes
_aws.CloudWatchMetrics.0.Metrics.4.Name | tmp_used
_aws.CloudWatchMetrics.0.Metrics.4.Unit | Bytes
_aws.CloudWatchMetrics.0.Metrics.5.Name | memory_utilization
_aws.CloudWatchMetrics.0.Metrics.5.Unit | Percent
_aws.CloudWatchMetrics.0.Metrics.6.Name | total_memory
_aws.CloudWatchMetrics.0.Metrics.6.Unit | Megabytes
_aws.CloudWatchMetrics.0.Metrics.7.Name | used_memory_max
_aws.CloudWatchMetrics.0.Metrics.7.Unit | Megabytes
_aws.CloudWatchMetrics.0.Namespace | LambdaInsights
_aws.LambdaInsights.ShareTelemetry | 1
_aws.Timestamp | 1678951430387
agent_memory_avg | 11
agent_memory_max | 11
agent_version | 1.0.143.0
billed_duration | 120000
billed_mb_ms | 122880000
cold_start | 0
cpu_system_time | 690
cpu_total_time | 1270
cpu_user_time | 580
duration | 120532
event_type | performance
fd_max | 1024
fd_use | 21
function_name | fhir-service-dev-fhirServer
memory_utilization | 41
request_id | 8fec99e8-fd41-458c-b8b6-c11f48c1b31e
rx_bytes | 1110
shutdown | 1
shutdown_reason | timeout
threads_max | 10
timeout | 1
tmp_free | 538324992
tmp_max | 550461440
tmp_used | 12136448
total_memory | 1024
total_network | 79305
tx_bytes | 78195
used_memory_max | 421
version | provisioned

</details>

Thank you.

**Comment by Bingjiling** [comment]
Hi @tanimura55 ,

Thanks for sharing the details! Based on the information you shared, this could be a Lambda timeout caused by DDB failed to response. Here're a few things you could do to tune the performance:

1. Deploy FWoA in a non-'dev' stage. When 'dev' is passed in as stage for FWoA deployment, some resources are created with less expensive configurations that is also less scalable. 
2. Check the DDB metrics, see if there's any extended delays during the time of load test. 
3. Update the Lambda timeout behavior. (reference: https://aws.amazon.com/premiumsupport/knowledge-center/lambda-function-retry-timeout-sdk/)
4. Use the latest FWoA version with CDK deployment. 
5. When running load test, configure a longer warm up period to allow resources to scale. 

Thanks,
Yanyu

**Comment by tanimura55** [comment]
Thanks for your kind comments, @Bingjiling !

So this load test may have reached DynamoDB's throughput limit, etc....
I will investigate DynamoDB and try some of the things you suggested as much as possible!

I'll comment as soon as I get some results, so just a moment(about a week), please. 

Thank you.

**Comment by tanimura55** [comment]
@Bingjiling 
Sorry for the late reply.

> 1. Deploy FWoA in a non-'dev' stage. When 'dev' is passed in as stage for FWoA deployment, some resources are created with less expensive configurations that is also less scalable.

Based on your feedback, it seems that the timeout was caused by the fact that the FHIR we tested was applied in a dev stage.
We have determined that it is probably not possible to avoid the problem in our test environment, so we are going to close our investigation of the cause.

If a similar event occurs in the production environment, we will investigate the DynamoDB metrics as well.

Thank you for your response.
This Issue is closed.
